### PR TITLE
Working background

### PR DESCRIPTION
**Pull request by Maximilian-Reuter**

### CODE REVIEW

File: src/main/cpp/input/InputHandler.h [hunk 1]
```diff
@@ -165,6 +167,8 @@ namespace nova {
 		struct mouse_button_event dequeue_mouse_button_event();
 		void queue_mouse_position_event(struct mouse_position_event  e);
 		struct mouse_position_event dequeue_mouse_position_event();
+        void queue_mouse_scroll_event(struct mouse_scroll_event  e);
```
Comment (DethRaid) [inline src/main/cpp/input/InputHandler.h]: You know you don't need to put `struct` here, right?

File: src/main/cpp/render/objects/gl_mesh.cpp [hunk 2]
```diff
@@ -15,8 +15,8 @@ namespace nova {
 
     gl_mesh::gl_mesh(mesh_definition &definition) {
         create();
-        set_data(definition.vertex_data, definition.vertex_format, usage::static_draw);
-        set_index_array(definition.indices, usage::static_draw);
+        set_data(definition.vertex_data, definition.vertex_format, usage::dynamic_draw);
```
Comment (DethRaid) [inline src/main/cpp/render/objects/gl_mesh.cpp]: Why?

File: src/main/java/com/continuum/nova/NovaNative.java [hunk 3]
```diff
@@ -11,6 +11,7 @@
     String GUI_ATLAS_NAME = "gui";
```
Comment (DethRaid) [inline src/main/java/com/continuum/nova/NovaNative.java]: Should probably make these static... that's probably my fault

File: src/main/java/com/continuum/nova/NovaRenderer.java [hunk 4]
```diff
@@ -68,7 +70,7 @@ public NovaRenderer() {
     public void onResourceManagerReload(@Nonnull IResourceManager resourceManager) {
         this.resourceManager = resourceManager;
 
-        if(firstLoad) {
+        if (firstLoad) {
```
Comment (DethRaid) [inline src/main/java/com/continuum/nova/NovaRenderer.java:73]: Reformatting boo

File: src/main/java/com/continuum/nova/NovaRenderer.java [hunk 5]
```diff
@@ -273,26 +282,32 @@ private void updateWindowSize(){
...
     }
 
     public void updateCameraAndRender(float renderPartialTicks, long systemNanoTime, Minecraft mc) {
-        if(NovaNative.INSTANCE.should_close()) {
+        if (NovaNative.INSTANCE.should_close()) {
             Minecraft.getMinecraft().shutdown();
         }
+
+        if (Minecraft.getMinecraft().currentScreen != null) {
+
```
Comment (DethRaid) [inline src/main/java/com/continuum/nova/NovaRenderer.java:303]: Why is there an empty if block here?

File: src/main/java/com/continuum/nova/NovaRenderer.java [hunk 6]
```diff
@@ -273,26 +282,32 @@ private void updateWindowSize(){
...
+        }
+        NovaDraw.novaDrawScreen(Minecraft.getMinecraft().currentScreen, renderPartialTicks);
+
         NovaNative.INSTANCE.execute_frame();
         updateWindowSize();
-        int scalefactor = new ScaledResolution(Minecraft.getMinecraft()).getScaleFactor()*2;
-        if (scalefactor!=this.scalefactor) {
+        int scalefactor = new ScaledResolution(Minecraft.getMinecraft()).getScaleFactor() * 2;
```
Comment (DethRaid) [inline src/main/java/com/continuum/nova/NovaRenderer.java:310]: The one time I like your reformatter

File: src/main/java/com/continuum/nova/gui/NovaDraw.java [hunk 7]
```diff
@@ -62,11 +76,12 @@ public static void draw(ResourceLocation texture, Integer[] indexBuffer, Float[]
      * @param vertices    the vertices as Vertex objects
      */
     public static void draw(ResourceLocation texture, Integer[] indexBuffer, Vertex[] vertices) {
+
         Float[] vertexbuffer = new Float[vertices.length * 9];
         for (int v = 0; v < vertices.length; v++) {
             vertexbuffer[v * 9] = vertices[v].x;
             vertexbuffer[v * 9 + 1] = vertices[v].y;
-            vertexbuffer[v * 9 + 2] = vertices[v].z;
+            vertexbuffer[v * 9 + 2] = currentZ;
```
Comment (DethRaid) [inline src/main/java/com/continuum/nova/gui/NovaDraw.java:84]: Since we're not getting the z from `Vertex` objects, let's cascade this change throughout NovaDraw

File: src/main/java/com/continuum/nova/gui/NovaDraw.java [hunk 8]
```diff
@@ -100,33 +111,41 @@ public static void draw(ResourceLocation texture, Integer[] indexBuffer, Vertex[
      * @param texWidth  texture / UV coordinates, relative to the original minecraft textures (not the texture atlas)
      * @param texHeight texture / UV coordinates, relative to the original minecraft textures (not the texture atlas)
      */
-    public static void drawRectangle(ResourceLocation texture, Rectangle2D.Float rect, float z, Rectangle2D.Float textureCoords) {
+    public static void drawRectangle(ResourceLocation texture, Rectangle2D.Float rect, float z, Rectangle2D.Float textureCoords,Color vertexColor) {
```
Comment (DethRaid) [inline src/main/java/com/continuum/nova/gui/NovaDraw.java]: No need to pass in Z here, since we're not using the passed in Z value anywhere

File: src/main/java/com/continuum/nova/gui/NovaDraw.java [hunk 9]
```diff
@@ -155,19 +174,25 @@ private static void computeCorrectMousePosition() {
...
+            NovaNative.INSTANCE.send_gui_buffer_command(command);
+            long end = System.nanoTime();
+            LOG.info("time used to copy buffers to c++ : " + (end - timePrev) + "time used to alloc buffers and fill: "+((end - timeWithAlloc) - (end - timePrev)));
+            Memory.purge();
         }
+
+
     }
 
     public static class Vertex {
```
Comment (DethRaid) [inline src/main/java/com/continuum/nova/gui/NovaDraw.java:194]: Please remove the z attribute from this class

**Review by DethRaid — CHANGES_REQUESTED** [review]
Please delete commented out code and cascade your changes so that `Vertex` doesn't have a z value

File: src/main/cpp/input/InputHandler.h (hunk 1 above)
Comment (Maximilian-Reuter) [reply src/main/cpp/input/InputHandler.h]: hmm okay was told that was neccessary

Comment (DethRaid) [reply src/main/cpp/input/InputHandler.h]: Nope! http://stackoverflow.com/questions/15181765/passing-structs-to-functions

In C, maybe

File: src/main/java/com/continuum/nova/NovaRenderer.java (hunk 4 above)
Comment (tobiasschulz) [reply src/main/java/com/continuum/nova/NovaRenderer.java:73]: why do you dislike these spaces so much? I can't think of anyone or any project that uses this particular code style...

Comment (DethRaid) [reply src/main/java/com/continuum/nova/NovaRenderer.java:73]: My code style is weird. I used to use the Vertical For Loop

```
for( int i = 0;
    i < x;
    i++ );
```

Then I moved on to super-spaced out everything

```
void foo( int x );
```

But I've since abandoned both of those.

I like not having spaces before parens in method calls and declarations and for/if/while statements is that this way it feels like the things inside the parens are more connected to the function/if/for/whatever. I recognize that this isn't what many people do, though, so I won't enforce this on your code. I booed on this line because it was my code, but had been reformatted. I'd rather we not reformat each other's code (and if I reformat yours, please call me out on it)

Comment (DethRaid) [reply src/main/java/com/continuum/nova/NovaRenderer.java:73]: (I don't see a way to edit that comment. Hopefully you get the idea)

File: src/main/java/com/continuum/nova/NovaNative.java [hunk 10]
```diff
@@ -11,7 +11,6 @@
...
     String BLOCK_COLOR_ATLAS_NAME = "block_color";
     String FONT_ATLAS_NAME = "font";
-    String OPTIONS_BACKGROUND = "options_background";
```
Comment (DethRaid) [inline src/main/java/com/continuum/nova/NovaNative.java]: Why was this line removed?

Comment (Maximilian-Reuter) [reply src/main/java/com/continuum/nova/NovaNative.java]: because it is now in free textures and does not need its own atlas anymore

File: src/main/java/com/continuum/nova/NovaRenderer.java (hunk 5 above)
Comment (Maximilian-Reuter) [reply src/main/java/com/continuum/nova/NovaRenderer.java:303]: it's not empty anymore

File: src/main/java/com/continuum/nova/NovaNative.java (hunk 3 above)
Comment (Maximilian-Reuter) [reply src/main/java/com/continuum/nova/NovaNative.java]: that would be redundant as it is an Interface

File: src/main/cpp/render/objects/gl_mesh.cpp (hunk 2 above)
Comment (Maximilian-Reuter) [reply src/main/cpp/render/objects/gl_mesh.cpp]: i wanted to find a way to reuse meshes and their buffers but didn't get anywhere. Reverted.

**Review by DethRaid — APPROVED** [review]
